firedrake/matrix.py

```diff
@@ -209,3 +251,57 @@ def __str__(self):
             ('' if self._assembled else 'un',
              self.a,
              self.bcs)
+
+    def force_evaluation(self):
+        "Ensures that the matrix is fully assembled."
+        self.assemble()
+        self._M._force_evaluation()
+
+
+class ImplicitMatrix(MatrixBase):
+    """A representation of the action of bilinear form operating
+    without explicitly assembling the associated matrix.  This class
+    wraps the relevant information for Python PETSc matrix.
+
+    :arg a: the bilinear form this :class:`Matrix` represents.
+
+    :arg bcs: an iterable of boundary conditions to apply to this
+        :class:`Matrix`.  May be `None` if there are no boundary
+        conditions to apply.
+
+
+    .. note::
+
+        This object acts to the right on an assembled :class:`.Function`
+        and to the left on an assembled cofunction (currently represented
+        by a :class:`.Function`).
+
+    """
+    def __init__(self, a, bcs, *args, **kwargs):
+        # sets self._a and self._bcs
+        super(ImplicitMatrix, self).__init__(a, bcs)
+
+        extra_ctx = kwargs.get("extra_ctx", {})
+
+        from firedrake.matrix_free.operators import ImplicitMatrixContext
+        ctx = ImplicitMatrixContext(a,
+                                    row_bcs=bcs,
+                                    col_bcs=bcs,
+                                    fc_params=kwargs["fc_params"],
+                                    context=extra_ctx)
+        self.petscmat = PETSc.Mat().create()
+        self.petscmat.setType("python")
+        self.petscmat.setSizes((ctx.row_sizes, ctx.col_sizes))
+        self.petscmat.setPythonContext(ctx)
+        self.petscmat.setUp()
+        self.petscmat.assemble()
+
+    def assemble(self):
+        self.petscmat.assemble()
```

Comment:
It is not idempotent.  It bumps the object state of the petsc matrix.  This has to be done so that preconditioners and solvers notice that the matrix has changed.